### ISSUE TITLE
bug in adding data to loop

### ISSUE THREAD

**Issue by kumar-physics** [uwbmrb/PyNMRSTAR#15]
I found a bug, while I am trying add data to loop. Here is how you can reproduce the bug

`import bmrb
ll=bmrb.Loop.from_scratch()
ll.add_column('_Atom_chem_shift.Auth_entity_assembly_ID')
ll.add_column('_Atom_chem_shift.Auth_seq_ID')
ll.add_column('_Atom_chem_shift.Auth_Comp_ID')
ll.add_column('_Atom_chem_shift.Auth_atom_ID')
ll.add_column('_Atom_chem_shift.Val')
ll.add_column('_Atom_chem_shift.Val_err')
ll.add_column('_Atom_chem_shift.Ambiguity_code')
dat1=['A','2','LYS','HB1','1.779323999994285','0.009758767141997112','2']
ll.add_data(dat1)
print ll
loop_
      _Atom_chem_shift.Auth_entity_assembly_ID
      _Atom_chem_shift.Auth_seq_ID
      _Atom_chem_shift.Auth_Comp_ID
      _Atom_chem_shift.Auth_atom_ID
      _Atom_chem_shift.Val
      _Atom_chem_shift.Val_err
      _Atom_chem_shift.Ambiguity_code

     A   2   LYS   HB1   1.779323999994285   0.009758767141997112   2    
   stop_

dat1[3]="HB2"
ll.add_data(dat1)
print ll
  loop_
      _Atom_chem_shift.Auth_entity_assembly_ID
      _Atom_chem_shift.Auth_seq_ID
      _Atom_chem_shift.Auth_Comp_ID
      _Atom_chem_shift.Auth_atom_ID
      _Atom_chem_shift.Val
      _Atom_chem_shift.Val_err
      _Atom_chem_shift.Ambiguity_code

     A   2   LYS   HB2   1.779323999994285   0.009758767141997112   2    
     A   2   LYS   HB2   1.779323999994285   0.009758767141997112   2    
   stop_
`

note HB1 in the first row become HB2, if I add HB3 then all 3 become HB3. Am I doing something wrong? or is it a bug?

**Comment by jonwedell** [comment]
It isn't a bug; it is just an artifact of the fact that python passes arguments by object reference. When you pass a list to a function you are in effect passing a pointer to the actual list such that any updates to the list are reflected in all places that have names bound to that list. For example:

```python
>>> a = [1,2,3]
>>> def test(a_list):
...   a_list.append(4)
... 
>>> test(a)
>>> a
[1, 2, 3, 4]
>>> test(a)
>>> a
[1, 2, 3, 4, 4]
```

[This page](http://www.python-course.eu/passing_arguments.php) has a great explanation of how it works.

Basically, in the code sample you provided above, you are in effect creating a loop whose data is a list of pointers to the list `dat1`. So when you modify `dat1` all of the rows change. To fix this just pass a shallow copy of the list as such:


```python
ll.add_data(dat1[:])
dat1[3]="HB2"
ll.add_data(dat1[:])
print ll
```

When you do this each row is pointing to a unique list which was copied from `dat1` - this way you can change dat1 without the data in the loop changing.

Jon
